service/dap/server.go

```diff
@@ -415,6 +424,42 @@ func (s *Server) handleRequest(request dap.Message) {
 				Body:     dap.ThreadsResponseBody{Threads: []dap.Thread{{Id: -1, Name: "Current"}}},
 			}
 			s.send(response)
+		case *dap.SetBreakpointsRequest:
+			// Halt running command
+			inProgress := s.debugger.RunningCommand()
+			// TODO(polina): once api.Call is run asynchronously, treat it like api.Continue as well
+			if inProgress == api.Continue {
+				s.overrideStopReason <- skipStop
+			} else {
+				// This would be one of the step commands, which get cancelled by debugger.
+				// We do not support any other running commands at this time, but if a new one
+				// gets added and this code doesn't get updated, the safest default behavior
+				// would be to declare the operation cancelled and let the user resume manually,
+				s.overrideStopReason <- fmt.Sprintf("cancelled %s", inProgress)
+			}
+			s.log.Debug("halting execution to set breakpoints")
+			_, err := s.debugger.Command(&api.DebuggerCommand{Name: api.Halt}, nil)
+			if err != nil {
+				s.sendErrorResponse(request.Request, UnableToSetBreakpoints, "Unable to set or clear breakpoints", err.Error())
+			}
+			// Set breakpoints
+			s.onSetBreakpointsRequest(request)
+			// Resume or cancel command
+			if inProgress == api.Continue {
```

Comment:
Is it guaranteed that s.overrideStopReason is drained by now?
Won't it ever happen that the previous running command stops for other reasons before the message in line 432/438 was queued to the channel?

Reply:
That's a good point. I guess it could happen in theory that things stop running after we check isRunning, but before we do all this. We wouldn't want to go back to the idea of locking sections of code with a running lock, so that any related processing is grouped with the state that triggers it, would we? To avoid blocking on next write and to avoid next doCommand acting on a stale channel value, I think drain it at the beginning of doCommand should be sufficient. What do you think?

Reply:
I think we could also run into this scenario:

1. setBreakpointsRequest checks inProgressCommand == api.Continue
2. program stops at breakpoint
3. checks override reason and finds none.
4. sends stopped event for breakpoint
5. setBreakpointsRequest send overrideReason = skipStop
5. setBreakpointsRequest attempts to halt program.
6. setBreakpointsRequest sets breakpoints
7. setBreakpointsRequest continues

The UI would show that the program is stopped at a breakpoint when it is really running.

Reply:
This reminds me of the analysis I did in async PR (https://github.com/go-delve/delve/pull/2423#discussion_r620454562) when trying to lock sections of code to avoid things happening like this out of order. At the time I could only think of a case where a misleading is-running error would arrive after a stopped event. But the idea was the same: if you decouple checking is-running and acting on it, you might end up carrying out while-running actions when things are no longer running. I think I got this sorted out with a dedicated lock in some intermediate version of the previous PR, following the right lock/unlock order relative to the resumeLoop channel to avoid deadlocks. But the overall multi-lock design made people uncomfortable. What is the alternative? Maybe use this skipStop channel for two-way communication, so handling breakpoints checks on what doCommand reported (i.e. if it acted on that signal) before sending a follow-up continue?

Reply:
I can take a closer look tomorrow.

Reply:
Is it too bad if setBreakpointsRequest triggers halt, and then fall through so it can be handled in the next loop like the other setBreakpointsRequests? Maybe the request can carry an extra info so after setBreakpointRequest is handled later, decide whether to continue or not depending on the stop reason. One drawback is the extra Stop event and Continue event, but will it be too visible for users?

Reply:
Extra Continue event would solve the issue of the UI and state being out of sync, but it won't solve the issue that we should not have continued in the first place because the stopped happened for different reasons.